v3/lints/rfc/lint_key_usage_and_extended_key_usage_inconsistent.go

```diff
@@ -0,0 +1,201 @@
+package rfc
+
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"fmt"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type KUAndEKUInconsistent struct{}
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_key_usage_and_extended_key_usage_inconsistent",
+		Description:   "The certificate MUST only be used for a purpose consistent with both key usage extension and extended key usage extension.",
+		Citation:      "RFC 5280, Section 4.2.1.12.",
+		Source:        lint.RFC5280,
+		EffectiveDate: util.RFC5280Date,
+		Lint:          NewKUAndEKUInconsistent,
+	})
+}
+
+func NewKUAndEKUInconsistent() lint.LintInterface {
+	return &KUAndEKUInconsistent{}
+}
+
+func (l *KUAndEKUInconsistent) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true when the certificate contains both a key usage
+// extension and an extended key usage extension.
+func (l *KUAndEKUInconsistent) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubscriberCert(c) && util.IsExtInCert(c, util.EkuSynOid) && util.IsExtInCert(c, util.KeyUsageOID)
+}
+
+// Execute returns an Error level lint.LintResult if the purposes of the certificate
+// being linted is not consistent with both extensions.
+func (l *KUAndEKUInconsistent) Execute(c *x509.Certificate) *lint.LintResult {
+	if len(c.ExtKeyUsage) > 1 {
+		return l.multiPurpose(c)
+	}
+	return l.strictPurpose(c)
+}
+
+// RFC 5280 4.2.1.12 on multiple purposes:
+//
+//	If multiple purposes are indicated the application need not recognize all purposes
+//	indicated, as long as the intended purpose is present.
+func (l *KUAndEKUInconsistent) multiPurpose(c *x509.Certificate) *lint.LintResult {
```

Comment:
This function has a rather high cyclomatic complexity, and I admit that my mind is having some trouble grokking it.

I'm going to continue to trace and verify it. However, if there are some additional code comments that can be added to explain the algorithm as well as the intended purpose of individual variables then that would be extremely appreciated.

Reply:
I have added some documentation to explain the algorithm